keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -3,703 +3,849 @@
+
+## Mutating Buckets
+
+As of the current design of COSI, mutating bucket properties is not supported. However, the current design does not prevent us from supporting it in the future. Mutable properties include encryption, object lifecycle configuration, replication configuration etc. These properties will be supported in future versions along with the capability to mutate them.
+
+These properties will be specified in the BucketRequest and follow the same pattern of events as Bucket creation. i.e. Bucket API object will be updated to reflect the properties set in BucketRequest, and then a controller will pick-up these changes and call the appropriate APIs to reflect them in the backend.
+
+# Object Lifecycle
+
+The following resources are managed by admins
+
+- BucketClass
+- BucketAccessClass
+
+The following resources are managed by a user. These are created with a reference to their corresponding class objects:
+
+- BucketClaim -> BucketClass
+- BucketAccess -> Bucket, BucketClaim, BucketAccessClass
+
+The COSI controller responds by creating the following objects
+
+- BucketRequest <-> Bucket (1:1)
+- BucketClaim -> Bucket
+
+Notes:
+
+ - There are **NO** cycles in the relationship graph of the above mentioned API objects.
+ - Mutations are not supported in the API.
 
-    rpc ProvisionerCreateBucket (ProvisionerCreateBucketRequest) returns (ProvisionerCreateBucketResponse) {}
-    rpc ProvisionerDeleteBucket (ProvisionerDeleteBucketRequest) returns (ProvisionerDeleteBucketResponse) {}
+When a user deletes the BucketRequest, then depending on the DeletionPolicy, the following happens:
 
-    rpc ProvisionerGrantBucketAccess (ProvisionerGrantBucketAccessRequest) returns (ProvisionerGrantBucketAccessResponse);
-    rpc ProvisionerRevokeBucketAccess (ProvisionerRevokeBucketAccessRequest) returns (ProvisionerRevokeBucketAccessResponse);
+- If deletionPolicy is Delete, then Bucket deletion is also triggered. 
+- If deletionPolicy is Retain, then Bucket is left as is, but the BucketRequest is deleted.
+
+Only when all accessors (BucketAccesses) of the Bucket are deleted, is the Bucket itself cleaned up. Having orphaned buckets in the cluster is an invalid state, unless the Bucket was imported into the cluster. 
+
+When a user deletes a BucketAccess, the corresponding secret/serviceaccount are also deleted. If a pod has that secret mounted when delete is called, then the secret will not be deleted, but will instead have its deletionTimestamp set. In this way, access to a Bucket is preserved until the application pod dies. 
```

Comment:
In the case of BucketAccess deletion while a pod is still running, does the access get revoked on the backend?